lib/engine/step/tracker.rb

```diff
@@ -73,7 +73,7 @@ def lay_tile(action, extra_cost: 0, entity: nil, spender: nil)
         free = false
         discount = 0
 
-        entity.abilities(:tile_lay) do |ability|
+        entity.abilities(:tile_lay, tile_ability ? tile_ability.when : nil) do |ability|
```

Comment:
I'll admit, this is kind of a hack. Let me explain how I got there and maybe someone else has a better idea.

The reason I need to change the code is because the special_track class supports the same private company having both a "when sold" track lay and an anytime track lay, which the E&K needs. However, the tracker module isn't aware of this and will try to use all the tile_laying abilities, which breaks the E&K. In summary, I need a way to leverage the special_track behavior without breaking the normal track step that also uses the tracker module.

The abilities method has its own hack, where if a single ability is found, it returns that and if multiple abilities are found, it returns an array. At first, I wrote code that supported those possibilities. In my opinion, it was quite messy and hard to read.

What makes the abilities return easier to handle is that it also yields, so you can simply add a block that handles both the single and multiple case...simple. I tried to leverage by writing a method in tracker called tile_lay_abilities, which took in the block and passed to the abilities method. I then overrode the method in the special_track class to provide the correct behavior for private company track lay abilities. It worked, but I worried about the tight coupling between this code, the new method, and the abilities method from the abilities module, so I abandoned in favor of what is in the code now.

Reply:
hm i think i favor using the yield as that was designed to handle multiple abilities of the same type, could i see what that looks like?

Reply:
Here is the diff.

diff --git a/lib/engine/step/g_1818/track.rb b/lib/engine/step/g_1817/track.rb
index 19b7520a..6d868a25 100644
--- a/lib/engine/step/g_1817/track.rb
+++ b/lib/engine/step/g_1817/track.rb
@@ -15,7 +15,7 @@ module Engine
           @hex = nil
         end
 
-        def lay_tile(action, extra_cost: 0, entity: nil, tile_ability: nil)
+        def lay_tile(action, extra_cost: 0, entity: nil)
           @game.game_error('Cannot lay and upgrade the same tile in the same turn') if action.hex == @hex
           @game.game_error('Cannot upgrade mines') if action.hex.assigned?('mine')
           super
diff --git a/lib/engine/step/g_1882/track.rb b/lib/engine/step/g_1882/track.rb
index cf9579a2..ccffdc74 100644
--- a/lib/engine/step/g_1882/track.rb
+++ b/lib/engine/step/g_1882/track.rb
@@ -10,7 +10,7 @@ module Engine
       class Track < Track
         include NwrTrackBonus
 
-        def lay_tile(action, extra_cost: 0, entity: nil, tile_ability: nil)
+        def lay_tile(action, extra_cost: 0, entity: nil)
           entity ||= action.entity
           super
           gain_nwr_bonus(action.tile, entity)
diff --git a/lib/engine/step/g_1889/special_track.rb b/lib/engine/step/g_1889/special_track.rb
index 78cd62c6..6d597cef 100644
--- a/lib/engine/step/g_1889/special_track.rb
+++ b/lib/engine/step/g_1889/special_track.rb
@@ -20,7 +20,7 @@ module Engine
           return super unless action.entity == @company
 
           lay_tile(action, spender: @round.company_seller)
-          ability(action.entity).use!
+          tile_lay_abilities(action.entity).use!
         end
 
         def process_pass(action)
diff --git a/lib/engine/step/g_18_ms/lay_tile_with_chattanooga_check.rb b/lib/engine/step/g_18_ms/lay_tile_with_chattanooga_check.rb
index 98f14a1f..2bde56f1 100644
--- a/lib/engine/step/g_18_ms/lay_tile_with_chattanooga_check.rb
+++ b/lib/engine/step/g_18_ms/lay_tile_with_chattanooga_check.rb
@@ -6,7 +6,7 @@ module Engine
       module LayTileWithChattanoogaCheck
         ACTION_HEX_OF_INTEREST = 'B10'
 
-        def lay_tile(action, extra_cost: 0, entity: nil, spender: nil, tile_ability: nil)
+        def lay_tile(action, extra_cost: 0, entity: nil, spender: nil)
           super
 
           entity ||= action.entity
diff --git a/lib/engine/step/special_track.rb b/lib/engine/step/special_track.rb
index bc6dc610..7f9cca8e 100644
--- a/lib/engine/step/special_track.rb
+++ b/lib/engine/step/special_track.rb
@@ -11,7 +11,7 @@ module Engine
       ACTIONS = %w[lay_tile].freeze
 
       def actions(entity)
-        return [] unless ability(entity)
+        return [] unless tile_lay_abilities(entity)
 
         ACTIONS
       end
@@ -25,25 +25,26 @@ module Engine
       end
 
       def blocks?
-        ability(@company)&.blocks
+        tile_lay_abilities(@company)&.blocks
       end
 
       def process_lay_tile(action)
-        tile_ability = ability(action.entity)
-        lay_tile(action, spender: action.entity.owner, tile_ability: tile_ability)
-        check_connect(action, tile_ability)
-        tile_ability.use!
+        ability = tile_lay_abilities(action.entity)
+        lay_tile(action, spender: action.entity.owner)
+        check_connect(action, ability)
+        ability.use!
       end
 
       def available_hex(entity, hex)
-        return if ability(entity)&.hexes&.any? && !ability(entity)&.hexes&.include?(hex.id)
-        return if ability(entity)&.reachable && !@game.graph.connected_hexes(entity.owner)[hex]
+        ability = tile_lay_abilities(entity)
+        return if ability&.hexes&.any? && !ability&.hexes&.include?(hex.id)
+        return if ability&.reachable && !@game.graph.connected_hexes(entity.owner)[hex]
 
         @game.hex_by_id(hex.id).neighbors.keys
       end
 
       def potential_tiles(entity, hex)
-        return [] unless (tile_ability = ability(entity))
+        return [] unless (tile_ability = tile_lay_abilities(entity))
 
         tiles = tile_ability.tiles.map { |name| @game.tiles.find { |t| t.name == name } }
         tiles = @game.tiles.uniq(&:name) if tile_ability.tiles.empty?
@@ -53,12 +54,12 @@ module Engine
           .select { |t| @game.phase.tiles.include?(t.color) && @game.upgrades_to?(hex.tile, t, tile_ability.special) }
       end
 
-      def ability(entity)
+      def tile_lay_abilities(entity, &block)
         return unless entity&.company?
 
-        ability = entity.abilities(:tile_lay, 'sold') if @round.respond_to?(:just_sold_company) &&
+        ability = entity.abilities(:tile_lay, 'sold', &block) if @round.respond_to?(:just_sold_company) &&
           entity == @round.just_sold_company
-        ability || entity.abilities(:tile_lay, 'track')
+        ability || entity.abilities(:tile_lay, 'track', &block)
       end
 
       def check_connect(action, tile_ability)
diff --git a/lib/engine/step/tracker.rb b/lib/engine/step/tracker.rb
index 353ea824..19589d66 100644
--- a/lib/engine/step/tracker.rb
+++ b/lib/engine/step/tracker.rb
@@ -38,7 +38,11 @@ module Engine
         @laid_track += 1
       end
 
-      def lay_tile(action, extra_cost: 0, entity: nil, spender: nil, tile_ability: nil)
+      def tile_lay_abilities(entity, &block)
+        entity.abilities(:tile_lay, &block)
+      end
+
+      def lay_tile(action, extra_cost: 0, entity: nil, spender: nil)
         entity ||= action.entity
         spender ||= entity
         tile = action.tile
@@ -73,7 +77,7 @@ module Engine
         free = false
         discount = 0
 
-        entity.abilities(:tile_lay, tile_ability ? tile_ability.when : nil) do |ability|
+        tile_lay_abilities(entity) do |ability|
           next if ability.hexes.any? && (!ability.hexes.include?(hex.id) || !ability.tiles.include?(tile.name))
 
           @game.game_error("Track laid must be connected to one of #{spender.id}'s stations") if ability.reachable &&

Reply:
Refactored to the block option.